src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,412 @@
+import edu.kit.compiler.optimizations.OptimizationState;
+import edu.kit.compiler.optimizations.Util;
+import firm.*;
+import firm.bindings.binding_irdom;
+import firm.bindings.binding_irgopt;
+import firm.bindings.binding_ircons.op_pin_state;
+import firm.bindings.binding_irnode.ir_opcode;
+import firm.bindings.binding_ircons.ir_cons_flags;
+import firm.bindings.binding_irnode;
+import firm.nodes.*;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.RequiredArgsConstructor;
+
+import java.util.*;
+
+import static firm.bindings.binding_irgraph.ir_graph_properties_t.IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE;
+
+@RequiredArgsConstructor
+@Data
+public class CommonSubexpressionElimination implements Optimization.Local {
+
+    private static final int MAX_INDIRECTIONS = 128;
+    private static final int MAX_CHANGES = 10;
+
+    // map that holds nodes that are to be exchanged with equivalent other nodes
+    private final HashMap<Node, Node> nodeValues = new HashMap<>();
+    private final Map<Pointer, Map<Pointer, Integer>> dominateMap = new HashMap<>();
+
+    @Override
+    public boolean optimize(Graph g, OptimizationState state) {
+        g.assureProperties(IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE);
+
+        boolean backEdgesEnabled = BackEdges.enabled(g);
+        if (!backEdgesEnabled) {
+            BackEdges.enable(g);
+        }
+
+        boolean changes;
+        int numChanges = 0;
+
+        do {
+            changes = false;
+            nodeValues.clear();
+            dominateMap.clear();
+
+            CSEVisitor visitor = new CSEVisitor(state);
+
+            var worklist = new StackWorklist<Node>(false);
+            g.walkTopological(new Util.NodeWorklistFiller(worklist));
+            while (!worklist.isEmpty()) {
+                worklist.dequeue().accept(visitor);
+            }
+
+            HashMap<Node, Node> replacementMap = new HashMap<>();
+            for (var replacement : nodeValues.entrySet()) {
+                // for each replacement merge nodes together
+                Node orig = findReplacement(replacement.getKey(), replacementMap);
+                Node node = findReplacement(replacement.getValue(), replacementMap);
+
+                if (orig != null && node != null) {
+                    assert orig.getOpCode() == node.getOpCode();
+                    changes |= transform(g, orig, node, orig.getOpCode(), replacementMap);
+                }
+            }
+            if (changes) {
+                numChanges++;
+            }
+
+            binding_irgopt.remove_bads(g.ptr);
+            binding_irgopt.remove_unreachable_code(g.ptr);
+            binding_irgopt.remove_bads(g.ptr);
+        } while (changes && numChanges < MAX_CHANGES);
+
+        if (!backEdgesEnabled) {
+            BackEdges.disable(g);
+        }
+
+        binding_irgopt.remove_bads(g.ptr);
+        binding_irgopt.remove_unreachable_code(g.ptr);
+        binding_irgopt.remove_bads(g.ptr);
```

Comment:
Is this not done before we leave the loop?

Reply:
Fair point